homeassistant/components/isy994/entity.py

```diff
@@ -0,0 +1,96 @@
+"""Representation of ISYEntity Types."""
+
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import Dict
+
+
+class ISYEntity(Entity):
+    """Representation of an ISY994 device."""
+
+    _attrs = {}
+    _name: str = None
+
+    def __init__(self, node) -> None:
+        """Initialize the insteon device."""
+        self._node = node
+        self._change_handler = None
+        self._control_handler = None
+
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to the node change events."""
+        self._change_handler = self._node.status.subscribe("changed", self.on_update)
+
+        if hasattr(self._node, "controlEvents"):
+            self._control_handler = self._node.controlEvents.subscribe(self.on_control)
+
+    def on_update(self, event: object) -> None:
+        """Handle the update event from the ISY994 Node."""
+        self.schedule_update_ha_state()
+
+    def on_control(self, event: object) -> None:
+        """Handle a control event from the ISY994 Node."""
+        self.hass.bus.fire(
```

Comment:
This is existing code, so just a note for the future. Firing custom events from entities is not according to our design guidelines. That should be done separately and should preferably not overlap with the state changed events that entities spawn.

Reply:
@shbatm 
We will likely need to have a long depreciation period for this as I know a lot of users rely on it for automations (https://www.home-assistant.io/integrations/isy994/#handling-insteon-control-events)

After we get through your PRs and this is converted to a config flow and it has devices, device Triggers might make sense as an alternative.

https://developers.home-assistant.io/docs/device_automation_trigger

Reply:
I implemented the original `on_control` events and am happy to be a part of the conversation and alternative solutioning. I'll do some reading on device automation triggers to see if they can completely cover the functionality.

Reply:
> After we get through your PRs and this is converted to a config flow and it has devices, device Triggers might make sense as an alternative.

@bdraco and @overloadut I agree, we can work through the most appropriate/"modern" Home Assistant methods and transition plan on the backside of these changes. 

This was core functionality for this integration  so it is relatively unchanged in these PRs except to add more detail to the events. 

The `controlEvent` subscription from PyISY still provides vital functionality for updating some entities' full state information (thermostats, multi-sensors, etc), but there may be better/newer ways of handling the event firing on the Home Assistant side.

Happy to follow-up and work with @bdraco and @overloadut after this set of updates.